PennMobile/Courses/CourseScheduleViewController.swift

```diff
@@ -56,18 +34,143 @@ extension CourseScheduleViewController {
+    override func viewDidDisappear(_ animated: Bool) {
+        super.viewDidDisappear(animated)
+        self.courseScheduleTableView.isHidden = true
+        self.courseScheduleTableView.refreshControl?.endRefreshing()
+        self.courseScheduleTableView.contentOffset = CGPoint.zero
+    }
+}
+
+// MARK: - Fetch and assign Model + Delegate
+extension CourseScheduleViewController {
+
+    private func fetchViewModel() {
+        if let courses = UserDefaults.standard.getCourses() {
+            assignModelAndRefreshTable(courses: courses)
+        } else {
+            if (Account.isLoggedIn) {
+                handleCourseRefresh()
+            } else {
+                self.showAlert(withMsg: "Please login to use this feature", title: "Login Error", completion: { self.navigationController?.popViewController(animated: true)} )
+            }
+        }
+    }
+
+    private func setCourseScheduleTableViewModel(_ model: CourseScheduleTableViewModel) {
+        self.model = model
+        self.model.delegate = self
+        courseScheduleTableView.model = self.model
+    }
+
+    private func assignModelAndRefreshTable(courses: Set<Course>) {
+        self.showActivity()
+        self.courseScheduleTableView.isHidden = false
+        model = try? CourseScheduleTableViewModel(courses: courses)
+        DispatchQueue.main.async {
+            self.setCourseScheduleTableViewModel(self.model)
+            self.courseScheduleTableView.reloadData()
+            self.hideActivity()
+        }
+    }
+}
+
+// MARK: - Networking
+extension CourseScheduleViewController {
+
+    func handleNetworkCourseRefreshResult(_ result: Result<Set<Course>, NetworkingError>) {
+        self.courseScheduleTableView.refreshControl?.endRefreshing()
+        switch result {
+        case .success(let courses):
+            self.showAlert(withMsg: "Your courses have been refreshed.", title: "Refresh Complete!", completion: { self.handleNetworkCourseRefreshCompletion(courses) })
+
+        case .failure(.noInternet):
+            self.showAlert(withMsg: "You appear to be offline.\nPlease try again later.", title: "Network Error", completion: { self.navigationController?.popViewController(animated: true) } )
+
+        case .failure(.parsingError):
+            self.showAlert(withMsg: "Penn's course servers are currently not updating. We hope this will be fixed shortly", title: "Uh oh!", completion: { self.navigationController?.popViewController(animated: true) })
+
+        case .failure(.authenticationError):
+            self.showAlert(withMsg: "Unable to access your courses.\nPlease login again.", title: "Login Error", completion: { self.handleAuthentication() })
+
+        default:
+            self.showAlert(withMsg: "Something went wrong.\nPlease try again later.", title: "Uh oh!", completion: { self.courseScheduleTableView.refreshControl?.endRefreshing(); self.navigationController?.popViewController(animated: true) } )
```

Comment:
Let's look into abstracting this error handling into a protocol, perhaps one that inherits from ShowsAlert (and thus must be a view controller). We should avoid duplicate code, especially when the error messages are the same.

Reply:
Got it. I'll work on it right away!